### PR TITLE
Add thread_time, process_time in `time` module

### PR DESCRIPTION
**Pull request by Snowapril**
This revision add four methods, `thread_time`, `thread_time_ns`, `process_time` and `process_time_ns`

### Documents
* https://docs.python.org/3.8/library/time.html#time.process_time
* https://docs.python.org/3.8/library/time.html#time.thread_time
* https://docs.rs/libc/0.2.92/libc/fn.clock_gettime.html
* https://docs.rs/libc/0.2.92/libc/constant.CLOCK_PROCESS_CPUTIME_ID.html
* https://docs.rs/libc/0.2.92/libc/constant.CLOCK_THREAD_CPUTIME_ID.html
* https://docs.rs/winapi/0.3.9/winapi/um/processthreadsapi/fn.GetProcessTimes.html
* https://docs.rs/winapi/0.3.9/winapi/um/processthreadsapi/fn.GetThreadTimes.html

### CODE REVIEW

**Comment by Snowapril** [comment]
I cant figure out getting process time and thread time in window platform without [`window-rs`](https://github.com/microsoft/windows-rs dependency),,, hmm..

**Comment by Snowapril** [comment]
Uhm.. maybe #3005 must be preceeded 😂

File: vm/src/stdlib/time.rs [hunk 1]
```diff
@@ -14,6 +14,16 @@ use crate::utils::Either;
...
 use crate::{PyClassImpl, PyObjectRef, PyResult, PyStructSequence, TryFromObject};
 
+#[cfg(windows)]
+use winapi::shared::{minwindef::FILETIME, ntdef::ULARGE_INTEGER};
+#[cfg(windows)]
+use winapi::um::processthreadsapi::{
+    GetCurrentProcess, GetCurrentThread, GetProcessTimes, GetThreadTimes,
+};
+
+#[cfg(not(windows))]
```
Comment (youknowone) [inline vm/src/stdlib/time.rs:24]:
```suggestion
#[cfg(not(any(windows, target_arch = "wasm32")))]
```

**Review by youknowone — CHANGES_REQUESTED** [review]
Yes, I think something broken by that. But wasm fail is not from that one. Would you fix this line and rebase it to upstream/main to see if it is fixed?

**Comment by youknowone** [comment]
Exotic platforms CI must be also checked. `CLOCK_PROCESS_CPUTIME_ID`, `CLOCK_THREAD_CPUTIME_ID` doesnt' exist in redox.

**Comment by Snowapril** [comment]
Yeah... I'm currently working on that issue. I found there are various cases where `clock_gettime` or `CLOCK_PROCESS_CPUTIME_ID` did not exist.

File: vm/src/stdlib/time.rs [hunk 2]
```diff
@@ -173,6 +197,169 @@ fn time_strptime(
...
+    };
+    let u = match tv.tv_usec.checked_mul(US_TO_NS) {
+        Some(value) => value,
+        None => {
+            return Err(vm.new_overflow_error("timestamp too large to convert to i64".to_owned()));
+        }
+    };
+
+    t.checked_add(u)
+        .ok_or_else(|| vm.new_overflow_error("timestamp too large to convert to i64".to_owned()))
```
Comment (youknowone) [inline vm/src/stdlib/time.rs:316]:
```suggestion
```suggestion
    (|tv| {
        let t = tv.tv_sec.checked_mul(SEC_TO_NS)?;
        let u = tv.tv_usec.checked_mul(US_TO_NS)?;
        t.checked_add(u)
    })(tv).ok_or_else(|| vm.new_overflow_error("timestamp too large to convert to i64".to_owned()))
```

I didnt' test the code but this kind of code will work

File: vm/src/stdlib/time.rs [hunk 3]
```diff
@@ -173,6 +197,169 @@ fn time_strptime(
...
+fn get_process_time(vm: &VirtualMachine) -> PyResult<Duration> {
+    cfg_if::cfg_if! {
+        if #[cfg(any(
+            target_os = "wasi",
+            target_os = "redox",
+            target_os = "illumos",
+            target_os = "netbsd",
+            target_os = "solaris"
+        ))] {
+            let mut ru: libc::rusage = std::mem::zeroed();
```
Comment (youknowone) [inline vm/src/stdlib/time.rs:329]:
```suggestion
            let mut ru: libc::rusage = unsafe { std::mem::zeroed() };
```
zeroed is unsafe

File: vm/src/stdlib/time.rs [hunk 4]
```diff
@@ -173,6 +197,169 @@ fn time_strptime(
...
 
+fn get_thread_time(vm: &VirtualMachine) -> PyResult<Duration> {
+    cfg_if::cfg_if! {
+        if #[cfg(windows)] {
+            let mut creation_time = FILETIME::default();
+            let mut exit_time = FILETIME::default();
+            let mut kernel_time = FILETIME::default();
+            let mut user_time = FILETIME::default();
+
+            unsafe {
```
Comment (youknowone) [inline vm/src/stdlib/time.rs:208]: this block looks like containing some safe codes

File: vm/src/stdlib/time.rs [hunk 5]
```diff
@@ -173,6 +197,169 @@ fn time_strptime(
...
+        }
+    }
+}
+
+fn time_thread_time(vm: &VirtualMachine) -> f64 {
+    match get_thread_time(vm) {
+        Ok(v) => v.as_secs_f64(),
+        Err(err) => panic!("Time error: {:?}", err),
+    }
+}
```
Comment (youknowone) [inline vm/src/stdlib/time.rs:253]: By looking inside `get_thread_time` the `Err` case looks able to be raised practically.

```suggestion
fn time_thread_time(vm: &VirtualMachine) -> PyResult<f64> {
    Ok(get_thread_time(vm)?.as_secs_f64())
}
```
And also other functions

File: vm/src/stdlib/time.rs [hunk 6]
```diff
@@ -173,6 +197,169 @@ fn time_strptime(
...
+        }
+
+        let mut large: ULARGE_INTEGER = std::mem::zeroed();
+        large.u_mut().LowPart = kernel_time.dwLowDateTime;
+        large.u_mut().HighPart = kernel_time.dwHighDateTime;
+        let ktime = *large.QuadPart();
+
+        large.u_mut().LowPart = user_time.dwLowDateTime;
+        large.u_mut().HighPart = kernel_time.dwHighDateTime;
+        let utime = *large.QuadPart();
```
Comment (youknowone) [inline vm/src/stdlib/time.rs:288]: it seems sharing some code with `get_thread_time`. how about adding a utility function for this case?

**Review by youknowone — CHANGES_REQUESTED** [review]

**Comment by youknowone** [comment]
because now we have lots of platform-specific code, defining a module can be convenient.

```rust
#[cfg(windows)]
mod windows {
   ...
}
```

File: vm/src/stdlib/time.rs (hunk 2 above)
Comment (Snowapril) [reply vm/src/stdlib/time.rs:316]: Oh great!! That way can avoid duplicates of error messages

File: vm/src/stdlib/time.rs (hunk 3 above)
Comment (Snowapril) [reply vm/src/stdlib/time.rs:329]: I missed that. Thanks. Is there any way to check grammar works on other platform?

File: vm/src/stdlib/time.rs (hunk 5 above)
Comment (Snowapril) [reply vm/src/stdlib/time.rs:253]: That way i can avoid panic call. Thanks

**Review by youknowone — APPROVED** [review]

File: vm/src/stdlib/time.rs [hunk 7]
```diff
@@ -173,6 +197,179 @@ fn time_strptime(
...
+fn time_thread_time(vm: &VirtualMachine) -> PyResult<f64> {
+    Ok(get_thread_time(vm)?.as_secs_f64())
+}
+
+fn time_thread_time_ns(vm: &VirtualMachine) -> PyResult<u64> {
+    Ok(get_thread_time(vm)?.as_nanos() as u64)
+}
+
+#[cfg(windows)]
+unsafe fn time_from_filtime(time: FILETIME) -> u64 {
```
Comment (youknowone) [inline vm/src/stdlib/time.rs:249]: I missed this part. `unsafe` keyword has different meaning for `unsafe fn` and `unsafe {}`. For `unsafe fn`, it means calling the fn is not safe. For `unsafe {}`, it mean the block is using unsafe operations. So in this case, each function calls are unsafe but the function itself is safe.

**Pull request merged**